client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java

```diff
@@ -244,41 +252,62 @@ private int receiveCommand(String commandId, Writer out, Writer err) {
-            ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
-            List<StreamType> streams = receiveResponse.getStream();
-            for (StreamType s : streams) {
-                byte[] value = s.getValue();
-                if (value == null) continue;
-                if (out != null && "stdout".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            out.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            out.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
-                    }
+
+            try {
+                numberOfReceiveCalls++;
+                ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
+                getStreams(receiveResponse, out, err);
+
+                System.out.println(String.format("receiveResponse: %s", receiveResponse.toString()));
+                CommandStateType state = receiveResponse.getCommandState();
+                if (COMMAND_STATE_DONE.equals(state.getState())) {
+                    return state.getExitCode().intValue();
                 }
-                if (err != null && "stderr".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            err.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            err.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
+            } catch (SOAPFaultException e) {
+                // e.getMessage().contains("The WS-Management service cannot complete the operation within the time specified in OperationTimeout")
+                if(e.getFault().getDetail().getFirstChild().getAttributes().getNamedItem("Code").getNodeValue().equals("2150858793")) {
+                    LOG.debug("Re-polling a receive request", this);
+                } else {
+                    throw e;
+                }
+            }
+        }
+    }
+
+    public int getNumberOfReceiveCalls() {
+        return numberOfReceiveCalls;
+    }
+
+    private void getStreams(ReceiveResponse receiveResponse, Writer out, Writer err) {
+        List<StreamType> streams = receiveResponse.getStream();
+        for (StreamType s : streams) {
+            byte[] value = s.getValue();
+            if (value == null) continue;
```

Comment:
Ok, I'm fine with not changing it just now.

It is *not* the same to check `value == null` outside the if block, because then it will never do the `s.isEnd()` check to close the stream.